design/open/hard-delete.md

```diff
@@ -0,0 +1,400 @@
+
+#### Reset / Delete Branch
+
+1. Perform reset / delete flow in the same manner as today
+2. On successful execution - create entries for all staging tokens under the `deleted` prefix
+
+### Background delete job
+
+* Scan committed tokens
+    * For each committed object - remove reference key
+    * For each staging token in committed entry - "Move staging token to deleted"   
+      All objects in committed staging tokens that were not actually committed are candidates for deletion, therefore
+  we can either execute the deleted tokens logic, or create entries for these tokens under the deleted prefix (implementation detail)
+    * Remove `staging/committed` key
+    * Delete staging area
+* Scan deleted tokens  
+  * For each token - remove references for all objects on staging token
+  * If it is the last reference for that object (list is empty after update) - perform hard-delete
+  * Delete staging area
+
+### Handling key override on same staging token - improvement
+
+TBD
+
+## 5. No Copies
+
+The idea is to make sure we do not enable copies, when no two entries in stage will point to the same physical object we will enable delete of the physical address in the following:
+
+- Upload - get previous entry and delete previous physical address unless staging token was updated
+- Revert - branch / prefix / object, when entry is dropped we can delete the physical address
+- Delete - repo/branch/object, delete the physical address after entry no longer exists
+
+We should enable the following to enable the physical delete:
+
+- Make sure we don't have the same physical address by sign links we return in our API. The logical path will be part of the signature that will be part of the physical address we generate. This will enable us to block and use of a physical address outside the repo/branch/entry.
+- The S3 gateway `put` operation with the copy support will use the underlying adapter to copy the data. Will require from each adapter to support/emulate a 'copy' operation.
+* (optional) Enable 'move' API as alternative to 'copy' we have seen that moving objects from one location to another by copy+delete of metadata will enable easy support for lakeFSFS move. This can be implemented by marking the entry on staging as 'locked'. During the move operation - lock+copy+delete+unlock operation. In case of any failure, as we don't have transactions, we may keep two entries, or keep locked object without ever delete its physical address.
+
+TODO(Barak): list more cases and how it will be address in this solution.
+
+### Flows
+
+#### Move operation
+1. Get entry
+2. "Lock" entry (`SetIf`)
+3. Copy entry to new path on current staging token
+4. Delete old entry
+5. "Unlock" entry (new entry) 
+
+#### Upload Object
+
+1. Get entry
+2. "Lock" entry if exists (override scenario) (`SetIf`)
+3. Write blob
+4. Add staged entry (or update)
+5. Delete physical address if previously existed
```

Comment:
If the physical path is predetermined by the logical path and the staging token, why do we need to delete anything?

Reply:
It is not - it was never suggested to couple the logical and physical path in this proposal (#5). We only ensure we have a single logical address for each physical address